### PR TITLE
Issue 5440 - memberof is slow on update/fixup if there are several 'g…

### PR DESCRIPTION
**Pull request by tbordaz**
…roupattr'

Bug description:
        When there are several groupattr (attr_1, attr_2,..) in memberof config
        To fixup entry 'e1', memberof does an internal search
        "(|(attr_1=e1)(attr_2=e1)...(attr_n=e1))"
        This is not valid regarding membership relation and in
        addition it prevents the server to bypass the filter evaluation.

Fix description:
        To fixup an entry iterate several internal searches
        "(attr_1=e1)" , then "(attr_2=e1)", then "(attr_n=e1)"

relates: #5440

Reviewed by: Pierre Rogier

### CODE REVIEW

File: ldap/servers/plugins/memberof/memberof.c [hunk 1]
```diff
@@ -745,115 +745,94 @@ memberof_call_foreach_dn(Slapi_PBlock *pb __attribute__((unused)), Slapi_DN *sdn
...
-    while (be) {
-        PRBool do_suffix_search = PR_TRUE;
-        if (!all_backends) {
-            be = slapi_be_select(sdn);
-            if (be == NULL) {
-                break;
-            }
-        }
-        if ((base_sdn = (Slapi_DN *)slapi_be_getsuffix(be, 0)) == NULL) {
+        /*slapi_log_err(SLAPI_LOG_INFO, MEMBEROF_PLUGIN_SUBSYSTEM, " memberof_call_foreach_dn search \"%s\"\n", filter_str);*/
```
Comment (mreynolds389) [inline ldap/servers/plugins/memberof/memberof.c]: Should this logging line be removed or set to PLUGIN logging?

**Review by mreynolds389 — APPROVED** [review]
LGTM

File: ldap/servers/plugins/memberof/memberof.c [hunk 2]
```diff
@@ -745,115 +745,93 @@ memberof_call_foreach_dn(Slapi_PBlock *pb __attribute__((unused)), Slapi_DN *sdn
...
+                if (memberof_entry_in_scope(config, base_sdn)) {
+                    /* do nothing, entry scope is spanning
+                     * multiple suffixes, start at suffix */
+                } else if (config->entryScopes) {
+                    for (size_t i = 0; config->entryScopes[i]; i++) {
+                        if (slapi_sdn_issuffix(config->entryScopes[i], base_sdn)) {
+                            /* Search each include scope */
+                            slapi_search_internal_set_pb(search_pb, slapi_sdn_get_dn(config->entryScopes[i]),
+                                    LDAP_SCOPE_SUBTREE, filter_str, 0, 0, 0, 0,
+                                    memberof_get_plugin_id(), 0);
```
Comment (droideck) [inline ldap/servers/plugins/memberof/memberof.c]: Just to keep it consistent with the rest of the code, could you, please, fix the indentation here:

```
                            slapi_search_internal_set_pb(search_pb, slapi_sdn_get_dn(config->entryScopes[i]),
                                                         LDAP_SCOPE_SUBTREE, filter_str, 0, 0, 0, 0,
                                                         memberof_get_plugin_id(), 0);
```

File: ldap/servers/plugins/memberof/memberof.c [hunk 3]
```diff
@@ -745,115 +745,93 @@ memberof_call_foreach_dn(Slapi_PBlock *pb __attribute__((unused)), Slapi_DN *sdn
...
-        if (do_suffix_search) {
-            slapi_search_internal_set_pb(search_pb, slapi_sdn_get_dn(base_sdn),
-                                         LDAP_SCOPE_SUBTREE, filter_str, 0, 0, 0, 0, memberof_get_plugin_id(), 0);
-            slapi_search_internal_callback_pb(search_pb, callback_data, 0, callback, 0);
-            slapi_pblock_get(search_pb, SLAPI_PLUGIN_INTOP_RESULT, &rc);
-            if (rc != LDAP_SUCCESS) {
-                break;
+            if (do_suffix_search) {
+                slapi_search_internal_set_pb(search_pb, slapi_sdn_get_dn(base_sdn),
+                        LDAP_SCOPE_SUBTREE, filter_str, 0, 0, 0, 0, memberof_get_plugin_id(), 0);
```
Comment (droideck) [inline ldap/servers/plugins/memberof/memberof.c]: And here:

```
                slapi_search_internal_set_pb(search_pb, slapi_sdn_get_dn(base_sdn),
                                             LDAP_SCOPE_SUBTREE, filter_str, 0, 0, 0, 0, memberof_get_plugin_id(), 0);
```

**Review by droideck — APPROVED** [review]
Besides the minor indent, looks good to me!

**Review by progier389 — CHANGES_REQUESTED** [review]
Something looks wrong:
slapi_pblock_new is called in a loop but it is freeed only once 
 ( I am also wondereing whether the slapi_pblock_init to reset the pblock in the be loop is correct or not (there me also be a leak here (but not related to this fix))

**Comment by tbordaz** [comment]
> Something looks wrong: slapi_pblock_new is called in a loop but it is freeed only once ( I am also wondereing whether the slapi_pblock_init to reset the pblock in the be loop is correct or not (there me also be a leak here (but not related to this fix))

Nice catch @progier389 , I updated the PR and will run asan tests.

**Comment by mreynolds389** [comment]
@tbordaz - I had an adjustment to your fix because I found it hard to follow the logic of search_pb and how it was being freed.  I think this is cleaner and less error prone.  Thoughts?

```
diff --git a/ldap/servers/plugins/memberof/memberof.c b/ldap/servers/plugins/memberof/memberof.c
index 9a49c1fb6..98955c330 100644
--- a/ldap/servers/plugins/memberof/memberof.c
+++ b/ldap/servers/plugins/memberof/memberof.c
@@ -782,7 +782,6 @@ memberof_call_foreach_dn(Slapi_PBlock *pb __attribute__((unused)), Slapi_DN *sdn
                 }
             }
 
-            search_pb = slapi_pblock_new();
             if (config->entryScopes || config->entryScopeExcludeSubtrees) {
                 if (memberof_entry_in_scope(config, base_sdn)) {
                     /* do nothing, entry scope is spanning
@@ -791,45 +790,44 @@ memberof_call_foreach_dn(Slapi_PBlock *pb __attribute__((unused)), Slapi_DN *sdn
                     for (size_t i = 0; config->entryScopes[i]; i++) {
                         if (slapi_sdn_issuffix(config->entryScopes[i], base_sdn)) {
                             /* Search each include scope */
+                            search_pb = slapi_pblock_new();
                             slapi_search_internal_set_pb(search_pb, slapi_sdn_get_dn(config->entryScopes[i]),
                                                          LDAP_SCOPE_SUBTREE, filter_str, 0, 0, 0, 0,
                                                          memberof_get_plugin_id(), 0);
                             slapi_search_internal_callback_pb(search_pb, callback_data, 0, callback, 0);
+                            slapi_pblock_destroy(search_pb);
                             /* We already did the search for this backend, don't
                              * do it again when we fall through */
                             do_suffix_search = PR_FALSE;
                         }
                     }
                 } else if (!all_backends) {
-                    slapi_pblock_destroy(search_pb);
                     break;
                 } else {
                     /* its ok, goto the next backend */
                     be = slapi_get_next_backend(cookie);
-                    slapi_pblock_destroy(search_pb);
                     continue;
                 }
             }
 
             if (do_suffix_search) {
+                search_pb = slapi_pblock_new();
                 slapi_search_internal_set_pb(search_pb, slapi_sdn_get_dn(base_sdn),
                                              LDAP_SCOPE_SUBTREE, filter_str, 0, 0, 0, 0,
                                              memberof_get_plugin_id(), 0);
                 slapi_search_internal_callback_pb(search_pb, callback_data, 0, callback, 0);
                 slapi_pblock_get(search_pb, SLAPI_PLUGIN_INTOP_RESULT, &rc);
+                slapi_pblock_destroy(search_pb);
                 if (rc != LDAP_SUCCESS) {
-                    slapi_pblock_destroy(search_pb);
                     break;
                 }
             }
 
             if (!all_backends) {
-                slapi_pblock_destroy(search_pb);
                 break;
             }
 
             be = slapi_get_next_backend(cookie);
-            slapi_pblock_destroy(search_pb);
         }
         slapi_ch_free_string(&filter_str);
     }

```

**Comment by tbordaz** [comment]
> Something looks wrong: slapi_pblock_new is called in a loop but it is freeed only once ( I am also wondereing whether the slapi_pblock_init to reset the pblock in the be loop is correct or not (there me also be a leak here (but not related to this fix))

@progier389 , I changed the logic to get a balanced slapi_pblock_new/slapi_pblock_destroy. Previous patch was leaking but now this change was tested (asan and in production) without leaks.

**Review by progier389 — APPROVED** [review]
LGTM
